### DISCUSSION TITLE
Named export raising type error and causing a crash

### DISCUSSION THREAD

**Discussion by joaobzrr** [pmndrs/zustand#1930] [Q&A]
While trying to import `create` from zustand like so:

```ts
import { create } from "zustand";
```

I get the following error:

```
Module '"zustand"' has no exported member 'create'. Did you mean to use 'import create from "zustand"' instead?
```

I tried ignoring it and running the dev server anyway, but it crashed. This was the error I got:

```
TypeError: (0 , zustand__WEBPACK_IMPORTED_MODULE_1__.create) is not a function
```

If I instead import `create` like this:

```ts
import create from "zustand"
```

it does work though I get a warning message:

```
[DEPRECATED] Default export is deprecated. Instead use import { create } from 'zustand'.
```

The Zustand version I'm on is 4.3.9.

**Comment by dai-shi** [comment]
Sounds like your bundler picks ESM, while it expects CJS. I think someone reported a similar thing before.

**Reply by joaobzrr** [reply]
How can I get named imports working? Or at least, get rid of the "DEPRECATED" warning. Should I downgrade zustand?

**Reply by dai-shi** [reply]
There must be a proper bundler config. Which version of webpack are you using? If you could make a small reproduction, someone like @barelyhuman could help.

Downgrading can be a workaround in the mean time.

**Reply by barelyhuman** [reply]
@joaobzrr a repro would help , and yes you can downgrade as a quick workaround

**Reply by joaobzrr** [reply]
@barelyhuman I can't upload the repo as this is a proprietary project. I was reading other issues related to this, and I guess this has something to do with the bundler we're using not being properly configured? I'm not sure what our bundler is, though. We're using nx which I guess uses webpack under the hood? Can you be more specific with what you need to know or what files I would need to provide?

**Reply by barelyhuman** [reply]
If there’s a webpack config then that. You can also go through the package.json file of the problematic package to help with other versions. 

The tsconfig is also something that’d give us more information and finally if this import is specifically breaking in a file. Can you tell if it’s a test runner file or not? 

If it’s a test runner file then please also put down the test engine being used (jest/mocha/etc)

**Reply by joaobzrr** [reply]
As far as I can tell, there's no webpack.config.js. Maybe nx uses a default webpack configuration if one isn't provided? 

I didn't understand what you meant by "You can also go through the package.json file of the problematic package to help with other versions". Can you clarify?

The import fails in every file where we do `import { create } from "zustand"`. None of them are test runner files.

Here are the tsconfig files:

**tsconfig.base.json**
```json
{
  "compileOnSave": false,
  "compilerOptions": {
    "rootDir": ".",
    "sourceMap": true,
    "declaration": false,
    "moduleResolution": "node",
    "emitDecoratorMetadata": true,
    "experimentalDecorators": true,
    "importHelpers": true,
    "target": "es2015",
    "module": "esnext",
    "lib": ["es2017", "dom"],
    "skipLibCheck": true,
    "skipDefaultLibCheck": true,
    "baseUrl": ".",
    "paths": {
      "@monorepo-project/core/base-application-client": [
        "libs/core/base-application-client/src/index.ts"
      ],
      "@monorepo-project/core/base-application-microservices": [
        "libs/core/base-application-microservices/src/index.ts"
      ],
      "@monorepo-project/infrastructure/database": [
        "libs/infrastructure/database/src/index.ts"
      ],
      "@monorepo-project/infrastructure/http": [
        "libs/infrastructure/http/src/index.ts"
      ],
      "@monorepo-project/infrastructure/nats": [
        "libs/infrastructure/nats/src/index.ts"
      ],
      "@monorepo-project/interfaces": ["libs/interfaces/src/index.ts"],
      "@monorepo-project/utils/schedule-validation": [
        "libs/utils/schedule-validation/src/index.ts"
      ],
      "@monorepo-project/utils/microservice-validations": [
        "libs/utils/microservice-validations/src/index.ts"
      ]
    }
  },
  "ts-node": {
    "transpileOnly": true,
    "files": true,
    "require": ["tsconfig-paths/register"],
    "compilerOptions": {
      "emitDecoratorMetadata": true,
      "experimentalDecorators": true,
      "baseUrl": ".",
      "paths": {
        "@monorepo-project/infrastructure/database": [
          "./libs/infrastructure/database/src/index.ts"
        ]
      }
    }
  },
  "exclude": ["node_modules", "tmp"]
}
```

**tsconfig.json**
```json
{
  "extends": "../../tsconfig.base.json",
  "compilerOptions": {
    "jsx": "preserve",
    "jsxImportSource": "@emotion/react",
    "allowJs": true,
    "esModuleInterop": true,
    "allowSyntheticDefaultImports": true,
    "strict": false,
    "forceConsistentCasingInFileNames": true,
    "noEmit": true,
    "resolveJsonModule": true,
    "isolatedModules": true,
    "incremental": true,
    "types": ["jest", "node"],
    "baseUrl": ".",
    "paths": {
      "@/*": ["./*"],
      "@assets/*": ["./assets/*"],
      "@components/*": ["./components/*"],
      "@contexts/*": ["./contexts/*"],
      "@hooks/*": ["./hooks/*"],
      "@infrastructure/*": ["./infrastructure/*"],
      "@modules/*": ["./modules/*"],
      "@pages/*": ["./pages/*"],
      "@utils/*": ["./utils/*"],
      "@types": ["./types"],
      "@monorepo-project/utils/microservice-validations": [
        "../../libs/utils/microservice-validations/src/index.ts"
      ],
      "@shared/interfaces": [
        "../../libs/interfaces/src/index.ts"
      ]
    }
  },
  "include": ["**/*.ts", "**/*.tsx", "**/*.js", "**/*.jsx", "next-env.d.ts"],
  "exclude": ["node_modules", "jest.config.ts"]
}
```

**Reply by joaobzrr** [reply]
Okay, this project uses Next.js so I think I can write a custom webpack config in next.config.js. I tried what you suggested [here](https://github.com/pmndrs/zustand/issues/454#issuecomment-886046399) but it didn't work.

```js
const { withNx } = require('./with-nx.js');
var path = require('path');

const nextConfig = {
  nx: {
    svgr: false,
  },
  trailingSlash: true,
  experimental: {
    outputFileTracingRoot: path.join(__dirname, '../../'),
  },
  webpack: (config, options) => {
    config.module.rules.push({
      type: "javascript/auto",
      test: /\.mjs$/,
      use: [],
    });
    return config;
  }
};

module.exports = withNx(nextConfig);
```

**Reply by barelyhuman** [reply]
So , NX, NextJS and no use of test tools. 
Cool, one other thing, can you also mention the version of NextJS that's being used?

**Reply by joaobzrr** [reply]
@barelyhuman 13.1.1.

**Reply by barelyhuman** [reply]
@joaobzrr  so the solution you tried above was for webpack 4, won’t work with next 13 as it uses rspack and/or webpack 5. 
thanks for the details. I should be able to create a reproduction with the above and will let you know post that

**Reply by barelyhuman** [reply]
Hey, @joaobzrr 
So I wasn't able to get it to crash with the details you've given but here's a repro if you wish to add or remove anything from it and can cause it to crash, it'd help with debugging 

https://github.com/barelyhuman/zustand-disc-1930-repro

**Reply by joaobzrr** [reply]
In the meantime, what version should I downgrade to to get rid of the warning?

**Reply by barelyhuman** [reply]
The warning is harmless, but I'd recommend you pin the version in the package.json for zustand because if we end up removing the method in the future, you might accidentally update to it.
